javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll

```diff
@@ -165,15 +165,11 @@ module DataFlow {
     predicate hasLocationInfo(
       string filepath, int startline, int startcolumn, int endline, int endcolumn
     ) {
-      filepath = "" and
-      startline = 0 and
-      startcolumn = 0 and
-      endline = 0 and
-      endcolumn = 0
+      none()
     }
 
     /** Gets the file this data flow node comes from. */
-    File getFile() { hasLocationInfo(result.getAbsolutePath(), _, _, _, _) }
+    File getFile() { none() } // overridden in subclasses
```

Comment:
Could we make the `Node` class and the `getFile` method abstract?

Reply:
No - that would be breaking for existing all subclasses of `Node`. While the change would probably be fine _internally_ in this library, it's mostly non-sense from an external point of view. Any subclass of `Node` would be seen as contributing a new kind of node in the IR, in most cases leading to negative recursion or re-evaluated IR code.

Making the predicate abstract would be less dramatic but still a compilation error for all existing subclasses.

This is why we prefer to use the range pattern instead of exposing abstract classes directly, but in the case of the `Node` class, you _can't_ actually contribute any new nodes because the underlying datatype is closed.